Service bindings docs (#3)
kabirsikand/cloudflare-docs-kabir: content/workers/platform/bindings/about-service-bindings.md

Service bindings allow you to:

* Segment multiple use cases into separate Workers Services that can be explicitly invoked from your code.
* Achieve better composability on the Workers platform using service-oriented architecture.
* Create private microservices, to be conditionally invoked from other edge-facing services.

While the interface among Service bindings is HTTP, the networking is not. Unlike the typical microservice architecture, where services communicate over a network and can suffer from latency or interruption, Service bindings are a zero-cost abstraction. When one Worker invokes another, there is no network delay and the request is executed immediately.

![Service bindings are a zero-cost abstraction](../media/service-bindings-comparison.png)

> ## Setting a Service binding
> You can manage Workers Service bindings in [**Workers**](https://dash.cloudflare.com/?zone=workers) > your **Worker** > **Settings**> **Variables** > **Service bindings** > **Edit variables**. You can also change the environment of a Workers Service binding, so you can target a specific version of a Workers Service.

Reviewer: ```suggestion
## Set a Service binding

To manage a Workers Service bindings:

1. Log in to the Cloudflare dashboard > Account Home > [Workers](https://dash.cloudflare.com/?zone=workers). 
2. Select your **Worker**.
3. Go to **Settings**> **Variables** > **Service bindings** > **Edit variables**. 

You can also change the environment of a Workers Service binding, so you can target a specific version of a Workers Service.
```

> ![To configure a Service binding, head to your Worker's Settings tab](../media/service-bindings-config.png)

Reviewer: ```suggestion
![To configure a Service binding, go to your Worker > Settings > Variables and follow the steps above](../media/service-bindings-config.png)
```

> Workers bound to your Worker will be listed in the [**Workers**](https://dash.cloudflare.com/?zone=workers) > your **Worker** > **Triggers** tab, under **Bound Services**. Your team can easily view cross-service dependencies in this manner.

Reviewer: ```suggestion
Workers bound to your Worker will be listed in [**Workers**](https://dash.cloudflare.com/?zone=workers) > your **Worker** > **Triggers** > **Bound Services**. Your team can easily view cross-service dependencies in this manner.
```

> ![Your team can view cross-service dependencies in the Triggers tab](../media/service-bindings-triggers.png)

Reviewer: ```suggestion
![Your team can view cross-service dependencies in the Cloudflare dashboard Account Home > Workers > Triggers](../media/service-bindings-triggers.png)
```

> ## Composing an example Worker

Reviewer: ```suggestion
## Compose an example Worker
```

> In the following example, we'll create a `gateway` Worker that invokes an `auth` Worker to handle authorization checks.

Reviewer: ```suggestion
In the following example, you will create a `gateway` Worker that invokes an `auth` Worker to handle authorization checks.
```

### Gateway Worker and Service bindings usage

> First, we'll create both a `gateway` and `auth` Worker. Once they're set up, navigate to the `gateway` Worker. In order to bind and call the [authentication](#authentication-workers-service) Worker, the `gateway` Worker needs to set up a Service binding. You can manage Service bindings in [**Workers**](https://dash.cloudflare.com/?zone=workers) > your **Worker** > **Settings**> **Variables** > **Service bindings** > **Edit variables**.

Reviewer: ```suggestion
First, you will create both a `gateway` and `auth` Worker. Once they are set up, go to the `gateway` Worker. In order to bind and call the [authentication](#authentication-workers-service) Worker, the `gateway` Worker needs to set up a Service binding. 

To manage Service bindings, go to Account Home > [**Workers**](https://dash.cloudflare.com/?zone=workers) > your **Worker** > **Settings**> **Variables** > **Service bindings** > **Edit variables**.
```

![Selecting Edit variables to create new bindings and edit existing bindings that enable Worker-to-Worker communication](../media/service-bindings.png)

Once added, the `gateway` Worker can access the Workers Service binding directly from the code, as in the example below. It utilizes the `fetch` API.

```js
export default {
  async fetch(request, env) {
    // Fetch AUTH service and pass request
    const authResponse = await env.auth.fetch(request.clone());

    // Return response from the AUTH service if the response status is not 200
    // It would return 403 'x-custom-token does not match, request not allowed' response in such case
    if (authResponse.status !== 200) {
      return authResponse;
    }

    // Request allowed
    // You can write application logic here
    // In this case we delegate the logic to an `application` Worker
    return await env.application.fetch(request)
  },
};
```

{{<Aside type="note">}}

Note that [Requests](../../runtime-apis/request) can only be read once. If you need to use a Request object multiple times, clone your incoming Request objects.

{{</Aside>}}

> In this setup, only the `gateway` Worker is exposed to the Internet and privately communicating with the `auth` and `application` Workers using Service bindings. Read on to build the `auth` Worker.

Reviewer: ```suggestion
In this setup, only the `gateway` Worker is exposed to the Internet and privately communicating with the `auth` and `application` Workers using Service bindings. In the next section, you will build the `auth` Worker.
```

### Authentication Workers Service

> The following authentication Workers Service code responds with `200` in the case that `x-custom-token` in the incoming request matches a `SECRET_TOKEN` Secret binding. Note that we implement `fetch` here, since a Service binding will invoke FetchEvent on the target Worker.

Reviewer: ```suggestion
The following authentication Workers Service code responds with a status code `200` in the case that `x-custom-token` in the incoming request matches a `SECRET_TOKEN` secret binding. Note that you implement `fetch` here, since a Service binding will invoke FetchEvent on the target Worker.
```

```js
export default {
  async fetch(request, env) {
    // Read x-custom-token header and make sure it matches SECRET_TOKEN
    if (request.headers.get('x-custom-token') === env.SECRET_TOKEN) {
      return new Response('Request allowed', { status: 200 });
    } else {
      return new Response('x-custom-token does not match, request not allowed', { status: 403 });
    }
  },
};
```

> This `auth` Workers Service does not need to have a `*.workers.dev` or other public endpoint; it is accessed through a Service binding from the `gateway` Worker directly. The authentication Worker is, effectively, a private Worker Service.

Reviewer: ```suggestion
This `auth` Workers Service does not need to have a `*.workers.dev` or other public endpoint. The `auth` Workers Service is accessed through a Service binding from the `gateway` Worker directly. The authentication Worker is, effectively, a private Worker Service.
```

> 

Reviewer: ```suggestion
```
